shared/chat/conversation/info-panel/menu/index.tsx

```diff
@@ -124,37 +160,77 @@ class InfoPanelMenu extends React.Component<Props> {
+          fullHeight={true}
+          fullWidth={true}
+          key="teamHeader"
+          style={Styles.collapseStyles([styles.channelHeader, styles.teamHeader])}
+        >
+          <Kb.Box2 direction="horizontal" gap="tiny">
+            <Kb.Avatar teamname={props.teamname} size={16} />
+            <Kb.Text type="BodyBold">{props.teamname}</Kb.Text>
+          </Kb.Box2>
+        </Kb.Box2>
+      ),
+    }
+
+    const hideItem = this.hideItem()
+    const muteItem = this.muteItem()
 
     const isAdhoc =
       (props.isSmallTeam && !props.convProps) || !!(props.convProps && props.convProps.teamType === 'adhoc')
-    const items: Kb.MenuItems = (isAdhoc
-      ? [
-          this.hideItem(),
-          this.muteItem(),
-          {danger: true, icon: 'iconfont-block-user', onClick: props.onBlockConv, title: 'Block'},
-        ]
-      : [
-          ...(props.canAddPeople ? addPeopleItems : []),
-          {
-            icon: 'iconfont-people',
-            onClick: props.onViewTeam,
-            style: {borderTopWidth: 0},
-            title: 'View team',
-          },
-          this.hideItem(),
-          this.muteItem(),
-          channelItem,
-          {danger: true, icon: 'iconfont-leave', onClick: props.onLeaveTeam, title: 'Leave team'},
-          {danger: true, icon: 'iconfont-remove', onClick: props.onBlockConv, title: 'Block team'},
-        ]
-    ).reduce<Kb.MenuItems>((arr, i) => {
-      i && arr.push(i as Kb.MenuItem)
-      return arr
-    }, [])
+    const items: Kb.MenuItems = []
+    if (isAdhoc) {
+      if (muteItem) {
+        items.push(muteItem as Kb.MenuItem)
+      }
+      if (hideItem) {
+        items.push(hideItem as Kb.MenuItem)
+      }
+      items.push({
+        danger: true,
+        icon: 'iconfont-block-user',
+        onClick: props.onBlockConv,
+        title: 'Block',
+      })
+    } else {
+      if (!props.isSmallTeam && !props.hasHeader) {
+        items.push(channelHeader)
+      }
+      if (muteItem) {
+        items.push(muteItem as Kb.MenuItem)
+      }
+      if (hideItem) {
+        items.push(hideItem as Kb.MenuItem)
+      }
+      if (!props.isSmallTeam && !props.isInChannel && !isGeneralChannel && !props.hasHeader) {
+        items.push({icon: 'iconfont-hash', onClick: props.onJoinChannel, title: 'Join channel'})
+      }
+      if (!props.isSmallTeam && props.isInChannel && !isGeneralChannel && !props.hasHeader) {
+        items.push({icon: 'iconfont-leave', onClick: props.onLeaveChannel, title: 'Leave channel'})
+      }
+      if (!props.isSmallTeam && !props.hasHeader) {
+        items.push(teamHeader)
+      }
+      items.push(channelItem, {
+        icon: 'iconfont-people',
+        onClick: props.onViewTeam,
+        title: 'Team info',
+      })
+      if (props.canAddPeople) {
+        addPeopleItems.forEach(item => items.push(item))
+      }
+      items.push({icon: 'iconfont-leave', onClick: props.onLeaveTeam, title: 'Leave team'})
+    }
```

Comment:
Lose the `else` to reduce the nesting

```ts
if (isAdhoc) {
    ...
    items.push({
        danger: true,
        icon: 'iconfont-block-user',
        onClick: props.onBlockConv,
        title: 'Block',
    })
    return
}

if (!props.isSmallTeam && !props.hasHeader) {
    // ...
}

// ...
```

Reply:
I think that would be useful if moving it to a separate function would clean things up, but that function would either need to have all the items preceding it passed in or created within it. So, either a messy function call or a the same mess further up/down the page. Unless I'm misunderstanding your feedback.